thriftpy2/protocol/apache_json.py

```diff
@@ -0,0 +1,309 @@
+            item_type = base_type[1]
+            if container_type == TType.STRUCT:
+                return self._dict_to_thrift(data, item_type)
+            elif container_type in (TType.LIST, TType.SET):
+                return [self._dict_to_thrift(v, item_type) for v in data[2:]]
+            elif container_type == TType.MAP:
+                return {
+                    self._dict_to_thrift(k, item_type[0]):
+                        self._dict_to_thrift(v, item_type[1]) for k, v in data[3].items()
+                }
+        result = {}
+        base_spec = base_type.thrift_spec
+        for field_idx, val in data.items():
+            thrift_spec = base_spec[int(field_idx)]
+            # spec has field type, field name, (sub spec), False
+            field_name = thrift_spec[1]
+            for ftype, value in val.items():
+                ttype = JTYPES[ftype]
+                if thrift_spec[0] == TType.BINARY and TType.BINARY != TType.STRING:
+                    bin_data = val.get('str', '')
+                    m = len(bin_data) % 4
+                    if m != 0:
+                        bin_data += '=' * (4-m)
+                    result[field_name] = base64.b64decode(bin_data)
+
+                elif ttype == TType.STRUCT:
+                    result[field_name] = self._dict_to_thrift(value, thrift_spec[2])
+                elif ttype in (TType.LIST, TType.SET):
+                    result[field_name] = [self._dict_to_thrift(v, thrift_spec[2]) for v in value[2:]]
+                elif ttype == TType.MAP:
+                    key_spec = thrift_spec[2][0]
+                    val_spec = thrift_spec[2][1]
+                    result[field_name] = {
+                        self._dict_to_thrift(k, key_spec): self._dict_to_thrift(v, val_spec)
+                        for k, v in value[3].items()
+                    }
+                else:
+                    result[field_name] = {
+                        'tf': bool,
+                        'i8': int,
+                        'i16': int,
+                        'i32': int,
+                        'i64': int,
+                        'dbl': float,
+                        'str': str,
+                    }[ftype](value)
+        if hasattr(base_type, '__call__'):
+            return base_type(**result)
+        else:
+            for k, v in result.items():
+                setattr(base_type, k, v)
+            return base_type
+
+    def read_struct(self, obj):
+        """
+        Read the next struct into obj, usually the argument from an incoming request
+        Only really used to read the arguments off a request into whatever we want
+        see thriftpy2.thrift.TProcessor.process_in for how this class will be used
+
+        Will turn the contents of self.req[4] into the args of obj,
+        ie. self.req[4]["1"] must be rendered into obj.thrift_spec
+
+        :param obj:
+        :return:
+        """
+        return self._dict_to_thrift(self._req[4], obj)
```

Comment:
it is better to add a new line at the end of the file?

Reply:
Done